index.js

```diff
@@ -45,72 +71,107 @@ function makeDots(num) {
+        const pinLength = this.props.pinLength ? this.props.pinLength : this.defaultProps.pinLength;
+        let pinCode = this.props.pinCode;
 
-        this.setState({value});
+        if (pinCode.length < pinLength) {
+            pinCode += String(num);
 
-        if (value.length == MAX_LENGTH) {
-            this.props.onSubmit(value);
-            this.props.onDone();
+            this.doCallback(pinCode);
         }
     }
 
-    handleRemove() {
-        const {value} = this.state;
-        this.setState({value: value.substr(0, value.length - 1)});
+    doCallback(pincode) {
+        this.props.onPinEntered(pincode);
     }
 
-    renderButton(num) {
-        return (<Text onPress={()=> this.handlePress(num)}
-                      style={styles.btn}>{num}
-        </Text>);
+    doClear() {
+        this.doCallback('');
     }
 
-    render() {
-        const {value} = this.state;
-        const marks = value.replace(/./g, ' ● ');
-        const dots = makeDots(MAX_LENGTH - value.length);
-
-        return (<View style={styles.pad} >
-            <Text style={styles.header} >
-                Enter pin:
-            </Text>
-
-            <View style={styles.row} >
-                <Text style={styles.pin} >{marks}{dots}</Text>
-            </View>
-
-            <View style={styles.row} >
-                {this.renderButton(1)}
-                {this.renderButton(2)}
-                {this.renderButton(3)}
-            </View>
+    doDelete() {
+        const pinCode = this.props.pinCode;
+        this.doCallback(pinCode.substr(0, pinCode.length - 1));
+    }
 
-            <View style={styles.row} >
-                {this.renderButton(4)}
-                {this.renderButton(5)}
-                {this.renderButton(6)}
-            </View>
+    renderButton(num) {
+        return (
+            <TouchableOpacity onPress={()=> this.handlePress(num)} style={styles.circle}>
+                <Text style={styles.btn}>{num}</Text>
+            </TouchableOpacity>
+        );
+    }
 
-            <View style={styles.row} >
-                {this.renderButton(7)}
-                {this.renderButton(8)}
-                {this.renderButton(9)}
-            </View>
+    buildButton(isVisible, text, clickHandler) {
+        if (isVisible) {
+            return <TouchableOpacity style={styles.invisibleCircle} onPress={clickHandler}><Text style={styles.clearBtn}>{text}</Text></TouchableOpacity>;
+        } else {
+            return <TouchableOpacity style={styles.invisibleCircle} />;
+        }
+    }
 
-            <View style={styles.row} >
-                <Text onPress={()=> this.handleClear()} style={styles.btn}>C</Text>
-                {this.renderButton(0)}
-                <Text onPress={()=> this.handleRemove()} style={styles.btn}>{'<'}</Text>
+    render() {
+        // Extract props
+        const pinCode = this.props.pinCode;
+        const pinLength = this.props.pinLength ? this.props.pinLength : this.defaultProps.pinLength;
```

Comment:
Since you fixed defaultProps defintion, you can simply do like this:

`
const {pinCode, pinLength, prompt} = this.props;
`